src/redeem_token.cc

```diff
@@ -177,75 +135,78 @@ void RedeemToken::OnCreateConfirmation(
+    OnRedeem(FAILED, unblinded_token);
     return;
   }
 
   // Parse JSON response
-  std::unique_ptr<base::Value> value(base::JSONReader::Read(response));
-  base::DictionaryValue* dictionary;
-  if (!value->GetAsDictionary(&dictionary)) {
-    BLOG(ERROR) << "Invalid response";
-    OnRedeem(FAILED);
+  std::unique_ptr<base::DictionaryValue> dictionary =
+      base::DictionaryValue::From(base::JSONReader::Read(response));
+
+  if (!dictionary) {
+    BLOG(ERROR) << "Failed to parse response: " << response;
+    OnRedeem(FAILED, unblinded_token);
     return;
   }
 
-  base::Value *v;
-
   // Get id
-  if (!(v = dictionary->FindKey("id"))) {
+  auto* id_value = dictionary->FindKey("id");
+  if (!id_value) {
     BLOG(ERROR) << "Response missing id";
-    OnRedeem(FAILED);
+    OnRedeem(FAILED, unblinded_token);
     return;
   }
 
-  auto id = v->GetString();
+  auto id = id_value->GetString();
 
   // Validate id
-  if (confirmation_id != id) {
-    BLOG(ERROR) << "Response confirmation id: " << confirmation_id
-        << " does not match id: " << id;
-    OnRedeem(FAILED);
+  if (id != confirmation_id) {
+    BLOG(ERROR) << "Response id: " << id << " does not match confirmation id: "
+        << confirmation_id;
+    OnRedeem(FAILED, unblinded_token);
     return;
   }
 
   // Fetch payment token
-  FetchPaymentToken(confirmation_id, payment_token_base64,
-      blinded_payment_token_base64);
+  FetchPaymentToken(confirmation_id, payment_token, blinded_payment_token,
+      unblinded_token);
 }
 
 void RedeemToken::FetchPaymentToken(
     const std::string& confirmation_id,
-    const std::string& payment_token_base64,
-    const std::string& blinded_payment_token_base64) {
+    const Token& payment_token,
+    const BlindedToken& blinded_payment_token,
+    const UnblindedToken& unblinded_token) {
+  DCHECK(!confirmation_id.empty());
+
   BLOG(INFO) << "FetchPaymentToken";
 
-  // GET /v1/confirmation/{confirmation_id}/paymentToken
   BLOG(INFO) << "GET /v1/confirmation/{confirmation_id}/paymentToken";
-
-  auto endpoint = std::string("/v1/confirmation/").append(confirmation_id)
-      .append("/paymentToken");
-  auto ads_serve_url = helper::AdsServe::GetURL().append(endpoint);
+  FetchPaymentTokenRequest request;
 
   BLOG(INFO) << "URL Request:";
-  BLOG(INFO) << "  URL: " << ads_serve_url;
+
+  auto url = request.BuildUrl(confirmation_id);
+  BLOG(INFO) << "  URL: " << url;
+
+  auto method = request.GetMethod();
 
   auto callback = std::bind(&RedeemToken::OnFetchPaymentToken,
-      this, ads_serve_url, _1, _2, _3, payment_token_base64,
-      blinded_payment_token_base64);
+      this, url, _1, _2, _3, payment_token, blinded_payment_token,
+      unblinded_token);
 
-  confirmations_client_->URLRequest(ads_serve_url, {}, "", "",
-      URLRequestMethod::GET, callback);
+  confirmations_client_->LoadURL(url, {}, "", "", method, callback);
 }
 
 void RedeemToken::OnFetchPaymentToken(
```

Comment:
this method is huge and should be broken up, but it also has no tests that I can see which is scary given how big it is

Reply:
This method is 16 lines long?

Reply:
am I misreading something? Starts at line 188 and ends at line 341?

Reply:
sometimes I miss collapsed code in the diff, but I looked again and don't see a method ending that I missed